### PR TITLE
Add setting for minimum string length to trigger auto popup

### LINKED ISSUE #105: Allow setting minimum string length for auto popup
I'd like to be able to set an option for the minimum length of a typed string before auto popup; if set to `3`, for instance, then the auto popup won't appear until I've typed at least three characters.

This could be configurable by source, as well.

### PR DESCRIPTION
**Pull request by jsit**
Closes #105

### CODE REVIEW

**Comment by prabirshrestha** [comment]
What about when the user types the trigger chars? Do we also want min length check then?
For example for `server.start`. Should the autocomplete trigger in `server.` or in `server.sta` if I have set min length to 3.

Should `force_refresh` ignore min length?

Could you also update docs.

File: autoload/asyncomplete.vim [hunk 1]
```diff
@@ -211,6 +211,12 @@ function! s:should_skip() abort
     endif
 endfunction
 
+function! s:get_min_chars(source_name) abort
+    let l:source = s:sources[a:source_name]
+    let l:min_chars = has_key(l:source, 'min_chars') ? l:source['min_chars'] : g:asyncomplete_min_chars
```
Comment (prabirshrestha) [inline autoload/asyncomplete.vim]: I think you should be able to do this directly

```vim
function! s:get_min_chars(source_name) abort
    return get(l:source, 'min_chars', g:asyncomplete_min_chars)
endif
```

Comment (jsit) [reply autoload/asyncomplete.vim]: Updated, thanks!

**Comment by prabirshrestha** [comment]
@jsit Any updates on this?

**Comment by jsit** [comment]
Looking into it, thanks!

**Comment by jsit** [comment]
> What about when the user types the trigger chars? Do we also want min length check then?
> For example for `server.start`. Should the autocomplete trigger in `server.` or in `server.sta` if I have set min length to 3.

I think the trigger char should always trigger auto-complete, regardless of length.

> Should `force_refresh` ignore min length?

I think so, yes.

**Comment by ranebrown** [comment]
@jsit @prabirshrestha Is this just waiting to get the docs updated before it gets merged or is there something else missing?

**Comment by jsit** [comment]
> @jsit @prabirshrestha Is this just waiting to get the docs updated before it gets merged or is there something else missing?

Adding documentation is a good idea, but other than that it just needs approval to be merged I believe.

**Comment by ranebrown** [comment]
It would be nice if this setting could be a buffer local variable so it would be possible to use different settings per filetype.

**Comment by Shatur95** [comment]
Very useful feature used in most IDEs. It would be great to see this in the master branch.

**Comment by prabirshrestha** [comment]
I think we still haven't solved the issue about force_refresh but I'm ok without this since I personally like min chars to be 1.

@jsit  Can you add docs and I should be able to merge this.

**Comment by jsit** [comment]
> What about when the user types the trigger chars? Do we also want min length check then?
> For example for `server.start`. Should the autocomplete trigger in `server.` or in `server.sta` if I have set min length to 3.

@prabirshrestha I think this will still popup with the trigger character, as I would expect it to (see [these lines](https://github.com/prabirshrestha/asyncomplete.vim/pull/135/files#diff-4948f573df0018e59dd56d5fcdcf473eR269-R279))

> Should `force_refresh` ignore min length?

@prabirshrestha Does `force_refresh` not force a refresh with these changes? I would expect it to still work. Not sure how to check this.

**Comment by jsit** [comment]
> It would be nice if this setting could be a buffer local variable so it would be possible to use different settings per filetype.

Your wish is my command! 🧙‍♂️

**Comment by Shatur95** [comment]
Is this branch is ready for merge?

**Comment by prabirshrestha** [comment]
Merged. Thanks

**Pull request merged**